### PR TITLE
Improve error messages with function name suggestion.

### PR DESCRIPTION
**Pull request by 2010YOUY01**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-datafusion/issues/6396

# Rationale for this change

When a wrong function name is given (possibly by typo, or missing underscore), suggest a similar valid function name:
```
DataFusion CLI v25.0.0
❯ select arrowtypeof();
Error during planning: Invalid function 'arrowtypeof'.
Did you mean 'arrow_typeof'?
```

# What changes are included in this PR?

When [binding a function](https://github.com/apache/arrow-datafusion/blob/3907997cd0bcf24e1f4e50c272ec50265bc07c23/datafusion/sql/src/expr/function.rs#L34), parser only knows is this function a window function or non-window function.
If it's non-window function, suggest the most similar function name from all `BuiltinScalarFunction` and `AggreagteFunction`
If it's window function, suggest the most similar function name from all `AggregateFunction` and `BuiltinWindowFunction`
Edit distance is used for the metric of similarity.

# Are these changes tested?
end-to-end sqllogictests are added.

# Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/sql/src/expr/function.rs [hunk 1]
```diff
@@ -84,73 +86,88 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
-                        aggregate_fun,
-                        function.args,
-                        schema,
-                        planner_context,
-                    )?;
-
-                    Expr::WindowFunction(expr::WindowFunction::new(
-                        WindowFunction::AggregateFunction(aggregate_fun),
-                        args,
+            if let Ok(fun) = self.find_window_func(&name) {
```
Comment (2010YOUY01) [inline datafusion/sql/src/expr/function.rs:89]: The diff display on GitHub is not clear, it's just a simple refactor
Before:
```rust
if let Some(WindowType::WindowSpec(window)) = function.over.take() {
		/* 1 */
		let fun = self.find_window_func(&name)?;
		/* 2 */
		return ...;
}

/* 3 */

// Return error
Err(...)
```
After:
```rust
if let Some(WindowType::WindowSpec(window)) = function.over.take() {
		/* 1 */
		if let Ok(fun) = self.find_window_func(&name) {
				/* 2 */
				return ...;
		}
} else {
		/* 3 */
}

// Return error
Err(...)
```

**Comment by alamb** [comment]
This looks awesome -- thank you @2010YOUY01  -- I plan to review this carefully tomorrow

File: datafusion/expr/Cargo.toml [hunk 2]
```diff
@@ -40,6 +40,7 @@ arrow = { workspace = true }
 datafusion-common = { path = "../common", version = "25.0.0" }
 lazy_static = { version = "^1.4.0" }
 sqlparser = "0.34"
+strsim = "0.10.0"
```
Comment (alamb) [inline datafusion/expr/Cargo.toml:43]: I reviewed this crate and it is widely used in the ecosystem (used by clap): https://crates.io/crates/strsim but hasn't been updated for 3 years

However, it does appear to be a new dependency for datafusion,

An alternate if we are worried about this new dependency is we could inline the definition of `levenshtein` into datafusion as it is not large

https://github.com/dguo/strsim-rs/blob/65eac453cbd10ba4e13273002c843e95c81ae93f/src/lib.rs#L192-L238

File: datafusion/expr/src/window_function.rs [hunk 3]
```diff
@@ -28,6 +28,7 @@ use arrow::datatypes::DataType;
 use datafusion_common::{DataFusionError, Result};
 use std::sync::Arc;
 use std::{fmt, str::FromStr};
+use strum_macros::EnumIter;
```
Comment (alamb) [inline datafusion/expr/src/window_function.rs:31]: TIL https://docs.rs/strum_macros/0.24.3/strum_macros/derive.EnumIter.html -- very cool

File: datafusion/sql/src/expr/function.rs (hunk 1 above)
Comment (alamb) [reply datafusion/sql/src/expr/function.rs:89]: The whitespace blind diff also shows this nicely: https://github.com/apache/arrow-datafusion/pull/6520/files?w=1

**Review by alamb — APPROVED** [review]
I tried this out locally and it was sooo cool!

```sql
DataFusion CLI v25.0.0
❯ select fooo('s');
Error during planning: Invalid function 'fooo'.
Did you mean 'floor'?
```

Really neat -- thank you @2010YOUY01  

The only question I have is about the new dependency (but it seems small and perhaps we could simply inline the implementation if we are worried about the implications of using it)

**Comment by alamb** [comment]
cc @comphead  / @mingmwang  / @mustafasrepo / @ozankabak  if you have any thoughts on the new dependency?

**Comment by ozankabak** [comment]
If we are only using Levenshtein, inlining sounds good to me as it is only a small piece of code. I suggest using the full package only if we can see use cases where we'd use it more extensively in the future.

**Comment by comphead** [comment]
> If we are only using Levenshtein, inlining sounds good to me as it is only a small piece of code. I suggest using the full package only if we can see use cases where we'd use it more extensively in the future.

Agree. the entire idea is neat to give a suggestion, however stale packages brings up not only future dependency conflicts but also unresolved security vulnerabilities. Perhaps we can replicate it from https://github.com/wooorm/levenshtein-rs ?

**Comment by 2010YOUY01** [comment]
Thank you all for the feedback! Makes sense, the `strsim` external dependency is replaced with inlining.

File: datafusion/common/src/utils.rs [hunk 4]
```diff
@@ -383,7 +377,7 @@ pub mod datafusion_strsim {
     /// required to change one string into the other.
     ///
     /// ```
-    /// use strsim::levenshtein;
+    /// use datafusion_common::utils::datafusion_strsim::levenshtein;
```
Comment (alamb) [inline datafusion/common/src/utils.rs:380]: 👍

File: datafusion/common/src/utils.rs [hunk 5]
```diff
@@ -318,6 +318,74 @@ impl<T: ?Sized> DataPtr for Arc<T> {
     }
 }
 
+/// Adopted from strsim-rs for string similarity metrics
+pub mod datafusion_strsim {
+    // Source: https://github.com/dguo/strsim-rs/blob/master/src/lib.rs
```
Comment (alamb) [inline datafusion/common/src/utils.rs:323]: 👍  thank you for the link

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @2010YOUY01

**Comment by alamb** [comment]
Thanks again @2010YOUY01  and @ozankabak

**Pull request merged**